lib/rtti.h

```diff
@@ -0,0 +1,297 @@
+/// (via ir-generator for IR::Node and friends) or provided explicitly by anyone willing
+/// to use lightweight RTTI. Two main components of RTTI are:
+///  - 64-bit typeids that could be either synthesized from class name at
+///    compile time or explicitly specified
+///  - Set of functions performing dynamic type checks and class hierarchy traversal
+///
+///  For example above in order to perform downcast to a virtual base we are
+///  doing "down-and-up" trick using virtual dispatch for `this` pointer adjustment. Briefly:
+///    - Inside `to<T>` we determine the typeid of IDeclaration
+///    - Call virtually toImpl, it lands with proper `this` inside
+///    - Declaration::toImpl; we know all immediate bases of Declaration via
+///      RTTI::TypeInfo<> due to RTTI annotation
+///    - From there we do upcast via TypeInfo::dyn_cast, finding the base type that corresponds
+///      to a typeId (of IDeclaration).
+///
+///  The implementation itself was crafted to allow as much compiler
+///  optimizations as possible. For example `is<T> / isA` after inlining
+///  compiles down to several integer comparisons (that are often combined into
+///  range checks). TypeInfo::dyn_cast looks expensive, however, everything is
+///  inlined and therefore is again series of typeid ranged checks interleaved
+///  with `this` adjustment.
+///  The code relies on C++17 features. Care was taken to ensure that `constexpr`
+///  functions are compiled down to compile-time constants for gcc >= 9 and recent
+///  of clang. Some of them likely would need to be marked as `consteval` when C++20
+///  would be used across the codebase.
+
+namespace RTTI {
+
+using TypeId = std::uint64_t;
+static constexpr TypeId InvalidTypeId = UINT64_C(0);
+
+namespace detail {
+// Apparently string_view does not optimize properly in constexpr context
+// for gcc < 13
+struct TypeNameHolder {
+    template <size_t N>
+    constexpr TypeNameHolder(const char (&str)[N]) : str(&str[0]), length(N - 1) {}  // NOLINT
+
+    const char *str;
+    size_t length;
+};
+
+// Derive the string containing the ultimate typename of `T`. The resulting
+// string should be available in constexpr context.
+// TODO: Probably use std::source_location::function_name with C++20
+template <typename T>
+constexpr TypeNameHolder typeNameProxy() {
+#ifdef __clang__
+    return __PRETTY_FUNCTION__;
+#elif defined(__GNUC__)
+    return __PRETTY_FUNCTION__;
+#else
+// TODO: We might use __FUNCSIG__ with MSVC should this will be necessary.
+#error "Unsupported compiler"
+#endif
+}
+
+static constexpr uint64_t FNV1a(const char *str, size_t n,
+                                uint64_t hash = UINT64_C(0xcbf29ce484222325)) {
+    return n == 0 ? hash : FNV1a(str + 1, n - 1, UINT64_C(0x100000001b3) * (hash ^ str[0]));
+}
+
+static constexpr uint64_t FNV1a(TypeNameHolder str) { return FNV1a(str.str, str.length); }
+
+// TypeIdResolver provides a unified way to specified type name `T`. Typeid
```

Comment:
"... provides a unified way to getting type id of the type `T`, or something like that might be better